src/crypto/CryptStateEVP.h

```diff
@@ -0,0 +1,280 @@
+// Copyright 2021 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+// This code aims to provide a unified interface to operate the EVP interface of OpenSSL.
+// See https://wiki.openssl.org/index.php/EVP_Authenticated_Encryption_and_Decryption
+
+#ifndef MUMBLE_CRYPTOSTATEEVP_H
+#define MUMBLE_CRYPTOSTATEEVP_H
+
+#include "CryptState.h"
+#include "EVPCiphers.h"
+
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+
+
+template< class T > class CryptStateEVP : public CryptState {
+public:
+	explicit CryptStateEVP();
+
+	bool isValid() const override;
+	void genKey() override;
+	bool setKey(const std::string &rkey, const std::string &eiv, const std::string &div) override;
```

Comment:
Isn't GCM supposed not to need an IV, or rather to calculate it from the key in a deterministic manner? Or does "IV" just mean "per-message nonce" here?

Reply:
Actually, GCM needs an IV. You can actually find a recommendation for IV generation in the link below:
https://github.com/mumble-voip/mumble/blob/d20d3e91ce08b7a5b11605c8a1aabcb083aadb5d/src/crypto/EVPCiphers.h#L14-L24

Reply:
Oh, so there's a separate API exposed somewhere that just takes the nonce and does the rest for us, like NaCl does?

Reply:
Yes. It is combined in the init function.

https://github.com/mumble-voip/mumble/blob/d20d3e91ce08b7a5b11605c8a1aabcb083aadb5d/src/crypto/CryptStateEVP.h#L138-L139